File: filly-core/src/widgets/range_slider.rs
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct RangeSliderWidget {
    pub title: String,
    pub min: i32,
    pub max: i32,
    pub value: i32,
    pub label: String,
    dirty: bool,
}

impl RangeSliderWidget {
    pub fn new(title: String, min: i32, max: i32, value: i32, label: String) -> Self {
        Self {
            title,
            min,
            max,
            value,
            label,
            dirty: true,
        }
    }
}

impl Widget for RangeSliderWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = 50u16.min(area.width.saturating_sub(2));
        let box_h = 5u16.min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let mut children = Vec::new();
        children.push(RenderNode::Text {
            rect: Rect::new(1, 0, box_w - 2, 1),
            content: self.title.clone(),
            alignment: Alignment::Center,
            style: TextStyle {
                bold: true,
                fg: Some("green".into()),
                ..TextStyle::normal()
            },
            role: None,
            label: None,
            description: None,
        });

        let range = self.max - self.min;
        let pos = if range == 0 {
            0
        } else {
            ((self.value - self.min) as f32 / range as f32 * (box_w - 4) as f32)
                as usize
        };
        let bar = format!(
            "{}{}{}",
            "─".repeat(pos),
            "●",
            "─".repeat((box_w - 4) as usize - pos)
        );
        children.push(RenderNode::Text {
            rect: Rect::new(2, 2, box_w - 4, 1),
            content: bar,
            alignment: Alignment::Left,
            style: TextStyle::accent(),
            role: None,
            label: None,
            description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, 3, box_w - 2, 1),
            content: format!("{}: {}", self.label, self.value),
            alignment: Alignment::Center,
            style: TextStyle::muted(),
            role: None,
            label: None,
            description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(),
            children,
            role: None,
            label: None,
            description: None,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match event {
            Event::Key(code, _) => match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: true,
                    error: None,
                }),
                KeyCode::Left => {
                    self.value = (self.value - 1).max(self.min);
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Right => {
                    self.value = (self.value + 1).min(self.max);
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Enter => EventResult::Response(WidgetResponse {
                    result: Some(serde_json::Value::Number(serde_json::Number::from(
                        self.value,
                    ))),
                    cancelled: false,
                    error: None,
                }),
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/range_slider.c
#include "range_slider.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

typedef struct {
    char *title;
    int min;
    int max;
    int value;
    char *label;
    bool dirty;
} RangeSliderData;

static void rs_render(Widget *self, Rect area, RenderTree *out) {
    RangeSliderData *d = (RangeSliderData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = 50, box_h = 7;
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    int range = d->max - d->min;
    int bar_w = box_w - 4;
    if (bar_w < 3) bar_w = 3;
    int pos = range == 0 ? 0 : (int)((float)(d->value - d->min) / range * (bar_w - 1));

    RenderTree *children = calloc(4, sizeof(RenderTree));
    children[0].type = RNODE_TEXT;
    children[0].rect = rect_new(1, 0, box_w - 2, 1);
    children[0].text.content = strdup(d->title);
    children[0].text.align = ALIGN_CENTER;
    children[0].text.style = textstyle_selected();

    children[1].type = RNODE_TEXT;
    children[1].rect = rect_new(1, 2, box_w - 2, 1);
    char *bar = calloc(bar_w + 1, 1);
    for (int i = 0; i < pos; i++) bar[i] = '-';
    bar[pos] = 'O';
    for (int i = pos + 1; i < bar_w; i++) bar[i] = '-';
    children[1].text.content = bar;
    children[1].text.align = ALIGN_LEFT;
    children[1].text.style = textstyle_accent();

    children[2].type = RNODE_TEXT;
    children[2].rect = rect_new(1, 4, box_w - 2, 1);
    char label[128];
    int pct = range == 0 ? 100 : (int)((float)(d->value - d->min) / range * 100);
    snprintf(label, sizeof(label), "%s: %d (%d%%)", d->label, d->value, pct);
    children[2].text.content = strdup(label);
    children[2].text.align = ALIGN_CENTER;
    children[2].text.style = textstyle_muted();

    children[3].type = RNODE_TEXT;
    children[3].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    children[3].text.content = strdup("Left/Right:adjust  Enter:confirm  Esc:cancel");
    children[3].text.align = ALIGN_CENTER;
    children[3].text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = 4;
}

static EventResult rs_handle_event(Widget *self, Event *ev, Backend *backend) {
    RangeSliderData *d = (RangeSliderData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_LEFT: d->value = d->value - 1 >= d->min ? d->value - 1 : d->min; d->dirty = true; return event_result_handled();
        case KEY_RIGHT: d->value = d->value + 1 <= d->max ? d->value + 1 : d->max; d->dirty = true; return event_result_handled();
        case KEY_ENTER:
            return event_result_response((WidgetResponse){ .result = cJSON_CreateNumber(d->value), .cancelled = false, .error = NULL });
        default: return event_result_unhandled();
    }
}

static bool rs_is_dirty(Widget *self) { return ((RangeSliderData *)(self + 1))->dirty; }
static void rs_clear_dirty(Widget *self) { ((RangeSliderData *)(self + 1))->dirty = false; }
static void rs_destroy(Widget *self) {
    RangeSliderData *d = (RangeSliderData *)(self + 1);
    free(d->title); free(d->label);
}

Widget *range_slider_widget_new(const char *title, int min, int max, int value, const char *label) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(RangeSliderData));
    w->vtable.render = rs_render;
    w->vtable.handle_event = rs_handle_event;
    w->vtable.is_dirty = rs_is_dirty;
    w->vtable.clear_dirty = rs_clear_dirty;
    w->vtable.destroy = rs_destroy;
    RangeSliderData *d = (RangeSliderData *)(w + 1);
    d->title = strdup(title);
    d->min = min; d->max = max; d->value = value;
    d->label = strdup(label);
    d->dirty = true;
    return w;
}